tests/federation_room_join_partial_state_test.go

```diff
@@ -3295,6 +3295,61 @@ func TestPartialStateJoin(t *testing.T) {
 		}
 	})
 
+	t.Run("Room stats are correctly updated once state re-sync completes", func(t *testing.T) {
+		deployment := Deploy(t, b.BlueprintAlice)
+		defer deployment.Destroy(t)
+
+		// create a user with admin powers as we will need this power to make the remote room visible in the
+		// local room list
+		terry := deployment.RegisterUser(t, "hs1", "terry", "pass", true)
+
+		server := createTestServer(t, deployment)
+		cancel := server.Listen()
+		defer cancel()
+		serverRoom := createTestRoom(t, server, terry.GetDefaultRoomVersion(t))
+
+		// start a partial state join
+		psjResult := beginPartialStateJoin(t, server, serverRoom, terry)
+		defer psjResult.Destroy(t)
+
+		// make the remote room visible in the local room list
+		reqBody := client.WithJSONBody(t, map[string]interface{}{
+			"visibility": "public",
+		})
+		terry.MustDoFunc(t, "PUT", []string{"_matrix", "client", "v3", "directory", "list", "room", serverRoom.RoomID}, reqBody)
+
+		// sanity check - before the state has completed syncing state we would expect only one user
+		// to show up in the room list
+		res := terry.MustDoFunc(t, "GET", []string{"_matrix", "client", "v3", "publicRooms"})
+		body, err := ioutil.ReadAll(res.Body)
+		if err != nil {
+			t.Fatalf("something broke: %v", err)
+		}
+		numJoinedMembers := gjson.GetBytes(body, "chunk.0.num_joined_members")
+		if numJoinedMembers.Int() != 1 {
+			t.Fatalf("Expected 1 users, returned %s", numJoinedMembers)
+		}
+
+		// finish syncing the state
+		psjResult.FinishStateRequest()
+		terry.MustSyncUntil(t,
+			client.SyncReq{},
+			client.SyncJoinedTo(terry.UserID, psjResult.ServerRoom.RoomID),
+		)
```

Comment:
Since Mat's PR to address non-lazy loading /syncs hasn't landed yet, /sync can return the room join before the room has been fully un-partial stated. Elsewhere we use:
```suggestion
		awaitPartialStateJoinCompletion(t, psjResult.ServerRoom, terry)
```
which blocks on /joined_members.

Reply:
This should address the random flakes, correct?

Reply:
Err actually it probably wouldn't, thinking about it for one second more.

Reply:
Good question. I think you're right in that it's not sufficient to address the flakes.